shapes/main.cpp

```diff
@@ -63,34 +77,46 @@ void findFirstShapeMatchingPredicate(const Collection& collection,
     }
 }
 
-constexpr int fibo(int n) {
-    if (n<=2) {
-        return 1;
-    } else {
-        return fibo(n - 1) + fibo(n - 2);
-    }
+int constexpr fibo(int number)
+{
+    if (number <= 2) { return 1; }
+    return fibo(number - 1) + fibo(number - 2);
+}
+
+template <typename T, typename = typename std::enable_if_t<std::is_base_of<Shape, T>::value>>
+void insertToCollection(Collection& collection, std::shared_ptr<T> shape) {
+    collection.push_back(shape);
+}
+
+template<class DerivedType, class... Arguments>
+std::shared_ptr<Shape> make_shape(Arguments&&... args) {
+    return make_shared<DerivedType>(decltype(args)(args)...);
```

Comment:
Hmm... you there is an additional redundant conversion with the way you implemented it.
It works, but the best way is to use perfect forwarding:
```suggestion
template<class DerivedType, class... Arguments>
std::shared_ptr<Shape> make_shape(Arguments&&... args) {
    return make_shared<DerivedType>(std::forward<Arguments>(args)...);
```

Reply:
done